File: oracle/src/aicad_oracle/v1/compound.py
"""Compound-curve expansion (SPEC-v1 §4.1 [D-19], [W0-6]), written from the SPEC member tables.

Inputs are evaluated values. Members are `<id>.<member>` in member-table order (rect: bottom,
c_br, right, c_tr, top, c_tl, left, c_bl; slot: right, cap_b, left, cap_a; polygon: e0 … e(n−1));
a member whose chord `|end − start|` is ≤ tol is omitted. All arcs are ccw. Invalid sizes raise
`CompoundError` with `INVALID_VALUE` (details `{field, value, expected}`), and a polygon `n`
that is not an exact integer `EXPR_NOT_INTEGER`, below 3 `INVALID_COUNT`.

A polygon's `n` may be as large as `MAX_COUNT_MAGNITUDE` (2^31). The oracle never loops that far:
when every side is at most tol/2 long, every member is omitted and the expansion is empty
(computed without the loop; exact, since the coordinates' rounding is far below tol there);
otherwise more than `MAX_POLYGON_SIDES` sides fail with the engine-internal
`ORACLE_RESOURCE_LIMIT` (ROBUSTNESS in the diff), never a hang.
"""

from __future__ import annotations

import math

from ..ir import Arc, Line
from .consts import LINEAR_TOLERANCE, MAX_COUNT_MAGNITUDE
from .expr import sin_cos_deg

TOL = LINEAR_TOLERANCE
#: The oracle's resource limit on polygon sides (not a SPEC rule; engine-internal code).
MAX_POLYGON_SIDES = 100_000


class CompoundError(Exception):
    def __init__(self, code: str, field: str, value: float, expected: str):
        super().__init__(f"{code}: {field} = {value}: expected {expected}")
        self.code = code
        self.field = field
        self.value = value
        self.expected = expected

    @property
    def details(self) -> dict:
        v = self.value if math.isfinite(self.value) else None
        return {"field": self.field, "value": v, "expected": self.expected}


def _bad(field: str, value: float, expected: str) -> CompoundError:
    return CompoundError("INVALID_VALUE", field, value, expected)


def _dist(a, b) -> float:
    return math.sqrt((b[0] - a[0]) * (b[0] - a[0]) + (b[1] - a[1]) * (b[1] - a[1]))


def _line(cid: str, member: str, s, e) -> Line | None:
    return Line(f"{cid}.{member}", (s[0], s[1]), (e[0], e[1])) if _dist(s, e) > TOL else None


def _arc(cid: str, member: str, c, s, e) -> Arc | None:
    return Arc(f"{cid}.{member}", (s[0], s[1]), (e[0], e[1]), (c[0], c[1]), True) if _dist(s, e) > TOL else None


def expand_rect(cid: str, center, corner, w: float, h: float, r: float) -> list:
    if not (math.isfinite(w) and w > TOL):
        raise _bad("w", w, f"> {TOL}")
    if not (math.isfinite(h) and h > TOL):
        raise _bad("h", h, f"> {TOL}")
    half_min = min(w, h) / 2.0
    if not (math.isfinite(r) and 0.0 <= r <= half_min):
        raise _bad("r", r, f"in [0, {half_min}] (min(w, h)/2)")
    if center is not None and corner is None:
        x0, x1, y0, y1 = center[0] - w / 2.0, center[0] + w / 2.0, center[1] - h / 2.0, center[1] + h / 2.0
    elif corner is not None and center is None:
        x0, x1, y0, y1 = corner[0], corner[0] + w, corner[1], corner[1] + h
    else:
        raise _bad("center", math.nan, "exactly one of center / corner")
    if r == 0.0:
        out = [_line(cid, "bottom", (x0, y0), (x1, y0)), _line(cid, "right", (x1, y0), (x1, y1)),
               _line(cid, "top", (x1, y1), (x0, y1)), _line(cid, "left", (x0, y1), (x0, y0))]
    else:
        out = [
            _line(cid, "bottom", (x0 + r, y0), (x1 - r, y0)),
            _arc(cid, "c_br", (x1 - r, y0 + r), (x1 - r, y0), (x1, y0 + r)),
            _line(cid, "right", (x1, y0 + r), (x1, y1 - r)),
            _arc(cid, "c_tr", (x1 - r, y1 - r), (x1, y1 - r), (x1 - r, y1)),
            _line(cid, "top", (x1 - r, y1), (x0 + r, y1)),
            _arc(cid, "c_tl", (x0 + r, y1 - r), (x0 + r, y1), (x0, y1 - r)),
            _line(cid, "left", (x0, y1 - r), (x0, y0 + r)),
            _arc(cid, "c_bl", (x0 + r, y0 + r), (x0, y0 + r), (x0 + r, y0)),
        ]
    return [m for m in out if m is not None]


def expand_slot(cid: str, a, b, w: float) -> list:
    length = _dist(a, b)
    if not (math.isfinite(length) and length > TOL):
        raise _bad("b", length, f"|b − a| > {TOL}")
    if not (math.isfinite(w) and w > TOL):
        raise _bad("w", w, f"> {TOL}")
    d = ((b[0] - a[0]) / length, (b[1] - a[1]) / length)
    m = (-d[1], d[0])
    h = w / 2.0
    am = (a[0] - h * m[0], a[1] - h * m[1])
    ap = (a[0] + h * m[0], a[1] + h * m[1])
    bm = (b[0] - h * m[0], b[1] - h * m[1])
    bp = (b[0] + h * m[0], b[1] + h * m[1])
    out = [_line(cid, "right", am, bm), _arc(cid, "cap_b", b, bm, bp),
           _line(cid, "left", bp, ap), _arc(cid, "cap_a", a, ap, am)]
    return [x for x in out if x is not None]


def expand_polygon(cid: str, center, n: float, size_field: str, size: float, rotation: float) -> list:
    if not (math.isfinite(n) and n == math.floor(n) and abs(n) <= MAX_COUNT_MAGNITUDE):
        raise CompoundError("EXPR_NOT_INTEGER", "n", n, "an exact integer")
    if n < 3.0:
        raise CompoundError("INVALID_COUNT", "n", n, ">= 3")
    if not (math.isfinite(size) and size > TOL):
        raise _bad(size_field, size, f"> {TOL}")
    if not math.isfinite(rotation):
        raise _bad("rotation", rotation, "finite")
    half = 180.0 / n
    sc = sin_cos_deg(half)
    if sc is None:
        raise _bad("n", n, "a finite polygon")
    s_half, c_half = sc
    if size_field == "circumradius":
        big_r = size
    elif size_field == "inradius":
        big_r = size / c_half
    elif size_field == "across_flats":
        big_r = size / (2.0 * c_half)
    else:
        big_r = size / (2.0 * s_half)
    count = int(n)
    if 2.0 * big_r * s_half <= 0.5 * TOL:
        return []  # every side ≤ tol/2: every member is omitted (§4.1)
    if count > MAX_POLYGON_SIDES:
        raise CompoundError("ORACLE_RESOURCE_LIMIT", "n", n,
                            f"<= {MAX_POLYGON_SIDES} sides (an oracle resource limit, not a SPEC rule)")
    pts = []
    for k in range(count):
        theta = rotation + (360.0 * float(k)) / n
        sn, cs = sin_cos_deg(theta)
        pts.append((center[0] + big_r * cs, center[1] + big_r * sn))
    out = [_line(cid, f"e{k}", pts[k], pts[(k + 1) % count]) for k in range(count)]
    return [x for x in out if x is not None]


def expand(c: dict, value) -> list:
    """Expand a compound curve dict whose Scalars are evaluated by `value(key_path_suffix, v)`.
    `value(v)` maps a Scalar to a float (literals pass through)."""
    k, cid = c["kind"], c["id"]
    if k == "rect":
        center = [value(x) for x in c["center"]] if "center" in c else None
        corner = [value(x) for x in c["corner"]] if "corner" in c else None
        return expand_rect(cid, center, corner, value(c["w"]), value(c["h"]), value(c.get("r", 0.0)))
    if k == "slot":
        return expand_slot(cid, [value(x) for x in c["a"]], [value(x) for x in c["b"]], value(c["w"]))
    assert k == "polygon"
    sizes = [f for f in ("circumradius", "inradius", "across_flats", "side") if f in c]
    return expand_polygon(cid, [value(x) for x in c["center"]], value(c["n"]), sizes[0],
                          value(c[sizes[0]]), value(c.get("rotation", 0.0)))

File: forge/crates/forge-ir/src/v1/compound.rs
//! Normative compound-curve expansion (SPEC-v1 §4.1 [D-19]): `rect`, `slot` and `polygon`
//! expand into member curves `<id>.<member>`, evaluated in f64 exactly as written; a member of
//! length ≤ tol is omitted ([W0-6]: a line's length is `|end − start|`, an arc's is its chord).
//!
//! Inputs are **evaluated** values (literals, or W1's evaluation of the expressions). W2 calls
//! this from sketch evaluation; W0's conformance suite checks it against
//! `corpus/v1/conformance/compound/`.

use super::LINEAR_TOLERANCE;
use super::degtrig::{cos_deg, sin_cos_deg, sin_deg};
use super::sketch::LiteralCurve;
use crate::P2;

/// A compound curve with evaluated fields.
#[derive(Debug, Clone, PartialEq)]
pub enum Compound {
    /// `center` xor `corner` (lower-left).
    Rect {
        center: Option<P2>,
        corner: Option<P2>,
        w: f64,
        h: f64,
        r: f64,
    },
    Slot {
        a: P2,
        b: P2,
        w: f64,
    },
    Polygon {
        center: P2,
        n: f64,
        size: PolygonSize,
        rotation: f64,
    },
}

/// The one size field of a polygon.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum PolygonSize {
    Circumradius(f64),
    Inradius(f64),
    AcrossFlats(f64),
    Side(f64),
}

impl PolygonSize {
    pub fn field(self) -> &'static str {
        match self {
            PolygonSize::Circumradius(_) => "circumradius",
            PolygonSize::Inradius(_) => "inradius",
            PolygonSize::AcrossFlats(_) => "across_flats",
            PolygonSize::Side(_) => "side",
        }
    }
    pub fn value(self) -> f64 {
        match self {
            PolygonSize::Circumradius(v)
            | PolygonSize::Inradius(v)
            | PolygonSize::AcrossFlats(v)
            | PolygonSize::Side(v) => v,
        }
    }
}

/// An invalid size (`INVALID_VALUE`, details `{ field, value, expected }`), or a count problem
/// (`INVALID_COUNT` / `EXPR_NOT_INTEGER` for a polygon's `n`).
#[derive(Debug, Clone, PartialEq, thiserror::Error)]
#[error("{code}: {field} = {value}: expected {expected}")]
pub struct CompoundError {
    pub code: &'static str,
    pub field: &'static str,
    pub value: f64,
    pub expected: String,
}

fn bad(field: &'static str, value: f64, expected: impl Into<String>) -> CompoundError {
    CompoundError {
        code: "INVALID_VALUE",
        field,
        value,
        expected: expected.into(),
    }
}

fn dist(a: P2, b: P2) -> f64 {
    ((b[0] - a[0]) * (b[0] - a[0]) + (b[1] - a[1]) * (b[1] - a[1])).sqrt()
}

fn line(id: &str, member: &str, start: P2, end: P2, construction: bool) -> Option<LiteralCurve> {
    (dist(start, end) > LINEAR_TOLERANCE).then(|| LiteralCurve::Line {
        id: format!("{id}.{member}"),
        start,
        end,
        construction,
    })
}

fn arc(
    id: &str,
    member: &str,
    center: P2,
    start: P2,
    end: P2,
    construction: bool,
) -> Option<LiteralCurve> {
    (dist(start, end) > LINEAR_TOLERANCE).then(|| LiteralCurve::Arc {
        id: format!("{id}.{member}"),
        start,
        end,
        center,
        ccw: true,
        construction,
    })
}

/// Expand a compound curve with id `id` into its member curves, in member-table order
/// (rect: bottom, c_br, right, c_tr, top, c_tl, left, c_bl; slot: right, cap_b, left, cap_a;
/// polygon: e0 … e(n−1)).
pub fn expand(
    id: &str,
    curve: &Compound,
    construction: bool,
) -> Result<Vec<LiteralCurve>, CompoundError> {
    let tol = LINEAR_TOLERANCE;
    let mut out = Vec::new();
    match *curve {
        Compound::Rect {
            center,
            corner,
            w,
            h,
            r,
        } => {
            if !(w.is_finite() && w > tol) {
                return Err(bad("w", w, format!("> {tol}")));
            }
            if !(h.is_finite() && h > tol) {
                return Err(bad("h", h, format!("> {tol}")));
            }
            let half_min = w.min(h) / 2.0;
            if !(r.is_finite() && r >= 0.0 && r <= half_min) {
                return Err(bad("r", r, format!("in [0, {half_min}] (min(w, h)/2)")));
            }
            let (x0, x1, y0, y1) = match (center, corner) {
                (Some(c), None) => (
                    c[0] - w / 2.0,
                    c[0] + w / 2.0,
                    c[1] - h / 2.0,
                    c[1] + h / 2.0,
                ),
                (None, Some(k)) => (k[0], k[0] + w, k[1], k[1] + h),
                _ => return Err(bad("center", f64::NAN, "exactly one of center / corner")),
            };
            let push = |out: &mut Vec<LiteralCurve>, c: Option<LiteralCurve>| out.extend(c);
            if r == 0.0 {
                push(
                    &mut out,
                    line(id, "bottom", [x0, y0], [x1, y0], construction),
                );
                push(
                    &mut out,
                    line(id, "right", [x1, y0], [x1, y1], construction),
                );
                push(&mut out, line(id, "top", [x1, y1], [x0, y1], construction));
                push(&mut out, line(id, "left", [x0, y1], [x0, y0], construction));
            } else {
                push(
                    &mut out,
                    line(id, "bottom", [x0 + r, y0], [x1 - r, y0], construction),
                );
                push(
                    &mut out,
                    arc(
                        id,
                        "c_br",
                        [x1 - r, y0 + r],
                        [x1 - r, y0],
                        [x1, y0 + r],
                        construction,
                    ),
                );
                push(
                    &mut out,
                    line(id, "right", [x1, y0 + r], [x1, y1 - r], construction),
                );
                push(
                    &mut out,
                    arc(
                        id,
                        "c_tr",
                        [x1 - r, y1 - r],
                        [x1, y1 - r],
                        [x1 - r, y1],
                        construction,
                    ),
                );
                push(
                    &mut out,
                    line(id, "top", [x1 - r, y1], [x0 + r, y1], construction),
                );
                push(
                    &mut out,
                    arc(
                        id,
                        "c_tl",
                        [x0 + r, y1 - r],
                        [x0 + r, y1],
                        [x0, y1 - r],
                        construction,
                    ),
                );
                push(
                    &mut out,
                    line(id, "left", [x0, y1 - r], [x0, y0 + r], construction),
                );
                push(
                    &mut out,
                    arc(
                        id,
                        "c_bl",
                        [x0 + r, y0 + r],
                        [x0, y0 + r],
                        [x0 + r, y0],
                        construction,
                    ),
                );
            }
        }
        Compound::Slot { a, b, w } => {
            let len = dist(a, b);
            if !(len.is_finite() && len > tol) {
                return Err(bad("b", len, format!("|b − a| > {tol}")));
            }
            if !(w.is_finite() && w > tol) {
                return Err(bad("w", w, format!("> {tol}")));
            }
            let d = [(b[0] - a[0]) / len, (b[1] - a[1]) / len];
            let m = [-d[1], d[0]];
            let h = w / 2.0;
            let a_minus = [a[0] - h * m[0], a[1] - h * m[1]];
            let a_plus = [a[0] + h * m[0], a[1] + h * m[1]];
            let b_minus = [b[0] - h * m[0], b[1] - h * m[1]];
            let b_plus = [b[0] + h * m[0], b[1] + h * m[1]];
            out.extend(line(id, "right", a_minus, b_minus, construction));
            out.extend(arc(id, "cap_b", b, b_minus, b_plus, construction));
            out.extend(line(id, "left", b_plus, a_plus, construction));
            out.extend(arc(id, "cap_a", a, a_plus, a_minus, construction));
        }
        Compound::Polygon {
            center,
            n,
            size,
            rotation,
        } => {
            if !n.is_finite() || n.fract() != 0.0 || n.abs() > super::MAX_COUNT_MAGNITUDE {
                return Err(CompoundError {
                    code: "EXPR_NOT_INTEGER",
                    field: "n",
                    value: n,
                    expected: "an exact integer".into(),
                });
            }
            if n < 3.0 {
                return Err(CompoundError {
                    code: "INVALID_COUNT",
                    field: "n",
                    value: n,
                    expected: ">= 3".into(),
                });
            }
            let s = size.value();
            if !(s.is_finite() && s > tol) {
                return Err(bad(size.field(), s, format!("> {tol}")));
            }
            if !rotation.is_finite() {
                return Err(bad("rotation", rotation, "finite"));
            }
            let half = 180.0 / n;
            let trig_err = || bad("n", n, "a finite polygon");
            let big_r = match size {
                PolygonSize::Circumradius(v) => v,
                PolygonSize::Inradius(v) => v / cos_deg(half).ok_or_else(trig_err)?,
                PolygonSize::AcrossFlats(v) => v / (2.0 * cos_deg(half).ok_or_else(trig_err)?),
                PolygonSize::Side(v) => v / (2.0 * sin_deg(half).ok_or_else(trig_err)?),
            };
            let count = n as usize;
            let mut pts = Vec::with_capacity(count);
            for k in 0..count {
                let theta = rotation + (360.0 * k as f64) / n;
                let (sn, cs) =
                    sin_cos_deg(theta).ok_or_else(|| bad("rotation", rotation, "finite"))?;
                pts.push([center[0] + big_r * cs, center[1] + big_r * sn]);
            }
            for k in 0..count {
                out.extend(line(
                    id,
                    &format!("e{k}"),
                    pts[k],
                    pts[(k + 1) % count],
                    construction,
                ));
            }
        }
    }
    Ok(out)
}

/// Every member suffix a compound kind can produce (for static `QUERY_UNKNOWN_CURVE` checks):
/// rect and slot members; for polygons `e0 … e(n−1)` (`n = None` when `n` is an expression:
/// every `e<k>` is then accepted).
pub fn member_names(kind: &str, n: Option<u32>) -> Vec<String> {
    match kind {
        "rect" => [
            "bottom", "c_br", "right", "c_tr", "top", "c_tl", "left", "c_bl",
        ]
        .map(String::from)
        .to_vec(),
        "slot" => ["right", "cap_b", "left", "cap_a"]
            .map(String::from)
            .to_vec(),
        "polygon" => (0..n.unwrap_or(0)).map(|k| format!("e{k}")).collect(),
        _ => Vec::new(),
    }
}
